### ISSUE TITLE
DFlash: child worker crashes on first decode (n_outputs_max family) — still broken after upstream sync

### ISSUE THREAD

**Issue by marksverdhei** [heiervang-technologies/ht-llama.cpp#108]
## Summary
The DFlash block-diffusion speculative decoder crashes on the **first decode**, and has since ~2026-05-27. Confirmed **still broken** after the 2026-06-13 upstream sync (`ht` @ `1aec77684`): upstream's `v_cells_impl` shared_ptr KV refactor — which mainlined ht's dual-context cell-sharing — did **not** fix it.

## Repro (titan deployment-shape smoke, 2026-06-13)
- Image: `zot.ht.local/unified-llm:stockui-1aec7768` (llama-server `9628/1aec77684`)
- Model: `gemma-4-31b-dflash-Q6_K`, `--spec-type dflash`
- Loads clean (slot + speculative ctx init succeed)
- **Child worker dies on the first decode** → router returns `500 "Failed to read connection"`
- Same failure family as the pre-sync `GGML_ASSERT n_outputs_max` on first decode

## Impact
- **Non-blocking for deploy.** The deployed speculative path is Gemma4 MTP (`--spec-type draft-mtp`), which is healthy post-sync (0.834 acceptance, dual-context KV active). DFlash is a separate experimental decoder.
- Crash is **isolated** — the router survived and MTP kept serving; only the DFlash child died.

## Likely area
First-decode output-buffer sizing (`n_outputs_max`) in the DFlash decode path — distinct from the KV cell-sharing the sync refactored, which is why the `v_cells_impl` change didn't address it.

_Evidence: titan post-sync rollout smoke (snoop-kube), 2026-06-13._

**Comment by marksverdhei** [comment]
## Root-cause diagnosis (code trace)

The crash is a draft-context **output-buffer under-sizing**, specific to DFlash. High confidence — the offending line is unconditional.

### The chain
1. The DFlash draft decodes the **entire diffusion block in one `llama_decode`, every token flagged `logits=true`** — `common/speculative.cpp:1325-1331`:
   ```cpp
   common_batch_clear(batch);
   for (int i = 0; i < block_size; i++) {            // block_size = dflash_block_size = 16 (llama-hparams.h:236)
       const llama_token tok = (i == 0) ? dp.id_last : mask_token_id;
       common_batch_add(batch, tok, n_ctx_used + i, { 0 }, /*logits=*/true);
   }
   llama_decode(ctx_dft_dec, batch);
   ```
   So `n_outputs_all == block_size == 16`.
2. `llama_decode` → `output_reserve(16)` → `llama-context.cpp:2254` computes `n_outputs_max = max(16, n_seq_max)` and `:2407` asserts `n_outputs_max <= cparams.n_outputs_max`.
3. **`cparams.n_outputs_max` for the draft context is `n_parallel` (1–4), not the block size.** Set in `tools/server/server-context.cpp:933`:
   ```cpp
   params_dft.n_outputs_max = params_base.n_parallel;   // unconditional — applies to the DFlash draft ctx
   ```
   `llama-context.cpp:209` then keeps it as-is (DFlash `has_encoder == false`, see `llama-model.cpp:2630`, so the `n_batch` override does NOT kick in).
4. `16 <= n_parallel` is false → `GGML_ASSERT` fires on the **first** `draft()` (first time the block is decoded). Matches the reported "loads clean, child dies on first decode".

### Why MTP survives but DFlash crashes
Line 933 sizes **all** draft contexts to `n_parallel`. MTP predicts `n_parallel` tokens per step (one per slot) → fits. DFlash decodes a whole `dflash_block_size` block per step → needs `n_outputs_max >= 16`. So the same line is correct for MTP and wrong for DFlash.

### Fix options
- **Preferred (in-library, robust):** in `llama_context` ctor (`llama-context.cpp:~209`), after computing `cparams.n_outputs_max`, clamp it up for DFlash:
  ```cpp
  if (model.arch == LLM_ARCH_DFLASH) {
      cparams.n_outputs_max = std::max(cparams.n_outputs_max, model.hparams.dflash_block_size);
  }
  ```
  `hparams` is loaded here, so `dflash_block_size` is available; this defends every caller, not just the server.
- **Server-side alternative:** at `server-context.cpp:933`, when the draft model is DFlash, set `params_dft.n_outputs_max = std::max(n_parallel, dflash_block_size)`. Caveat: block size must be read from the draft GGUF header (the model isn't fully loaded at 933, but `common_get_device_memory_data` at :949 already reads the header).

Either way the invariant is: **a DFlash decoder context must reserve `>= dflash_block_size` outputs.** A defensive grow-instead-of-assert in `output_reserve` would also turn the hard crash into a recoverable path.

*Diagnosis is a static code trace (the cluster repro wasn't re-run here); the root-cause line is unconditional so confidence is high, but worth a one-line confirm that `cparams.n_outputs_max < 16` in the failing child's logs.*
